### PR TITLE
Add an event to override XCGS_Unit::HasHeightAdvantageOver

### LINKED ISSUE #851: Add an event to override XCGS_Unit::HasHeightAdvantageOver
I'd like to put an event in the mentioned function so that my Jet Shot can benefit from various abilities that trigger if the shooter has a height advantage, such as Death From Above. After all, the whole point of that ability is to provide a height advantage for the shooter, which is currently done through a hit modifier.

Does `OverrideHasHeightAdvantageOver` work as the event name?

### PR DESCRIPTION
**Pull request by Iridar**
Fixes #851 

Already includes HL Docs. Tested in-game to work correctly.

### CODE REVIEW

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 1]
```diff
@@ -1108,12 +1108,85 @@ function bool EverHadTrait(name TraitTemplateName)
...
+/// whether one unit has height advantate over another unit.
+///
+///	In order to take advantage of this event, make a listener with the ELD_Immediate deferral, 
+/// and cast EventData to XComLWTuple. `Tuple.Data[2].b` will already contain the bool value
+///  of whether this unit has height advantage over the other unit according to the vanilla logic. 
+/// You can replace the value with your own ased on arbitrary parameters,
+/// such as one of the units being affected by a certain effect or having a certain ability.
+///	
+/// ```unrealscript
+/// EventID: OverrideUnitFocusUI
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:1136]:
```suggestion
/// EventID: OverrideHasHeightAdvantageOver
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 2]
```diff
@@ -1108,12 +1108,85 @@ function bool EverHadTrait(name TraitTemplateName)
...
+}
+//	Begin Issue #851
+/// HL-Docs: feature:OverrideHasHeightAdvantageOver; issue:851; tags:tactical
+/// This event allows mods to override the standard game's logic for checking
+/// whether one unit has height advantate over another unit.
+///
+///	In order to take advantage of this event, make a listener with the ELD_Immediate deferral, 
+/// and cast EventData to XComLWTuple. `Tuple.Data[2].b` will already contain the bool value
+///  of whether this unit has height advantage over the other unit according to the vanilla logic. 
+/// You can replace the value with your own ased on arbitrary parameters,
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:1132]: ased?

Comment (Iridar) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:1132]: based *

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc (hunk 1 above)
Comment (Iridar) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:1136]: whoops, thanks for the catch

**Comment by Iridar** [comment]
@Xymanek 

Please re-review this when you have a moment. It really should be good to go, but I don't wanna be skipping hoops on CHL additions.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 3]
```diff
@@ -1108,12 +1108,85 @@ function bool EverHadTrait(name TraitTemplateName)
...
+	Tuple.Data[1].kind = XComLWTVBool;
+	Tuple.Data[2].kind = XComLWTVBool;
+
+	Tuple.Data[0].o = OtherUnit;
+	Tuple.Data[1].b = bAsAttacker;
+	Tuple.Data[2].b = bHasHeightAdvantageOver;
+
+	`XEVENTMGR.TriggerEvent('OverrideHasHeightAdvantageOver', Tuple, self, none);
+
+	return Tuple.Data[2].b;
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:1187]: IMO it's better to have the return as `Data[0]`

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 4]
```diff
@@ -1108,12 +1108,85 @@ function bool EverHadTrait(name TraitTemplateName)
...
+/// You can replace the value with your own based on arbitrary parameters,
+/// such as one of the units being affected by a certain effect or having a certain ability.
+///	
+/// ```unrealscript
+/// EventID: OverrideHasHeightAdvantageOver
+/// EventData: XComLWTuple {
+///     Data: [
+///       inout XComGameState_Unit OtherUnit,
+///       inout bool bAsAttacker,
+///       inout bool HasHeightAdvantageOver
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:1141]: Docs mismatch - `OtherUnit` and `bAsAttacker` are `in`

Comment (Iridar) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc]: Fixed.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc (hunk 3 above)
Comment (Iridar) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc]: As you wish, fixed.

**Comment by Iridar** [comment]
Unfortunately I don't seem to be able to build the Highlander to test the changes requested by @Xymanek due to CanUnitMoveFromMelee issues that were fixed in the later versions of the Highlander, but I gave everything a long hard look and it should be good to go.

**Comment by Iridar** [comment]
Tested, everything works.

**Pull request merged**